gstreamer/src/query.rs

```diff
@@ -208,84 +208,273 @@ impl fmt::Debug for QueryRef {
+            ffi::GST_QUERY_ACCEPT_CAPS => QueryView::AcceptCaps(AcceptCaps(query)),
+            ffi::GST_QUERY_CAPS => QueryView::Caps(Caps(query)),
+            ffi::GST_QUERY_DRAIN => QueryView::Drain(Drain(query)),
+            ffi::GST_QUERY_CONTEXT => QueryView::Context(Context(query)),
+            _ => QueryView::Other(Other(query)),
         }
     }
 }
 
-pub enum QueryView<T> {
-    Position(Position<T>),
-    Duration(Duration<T>),
-    Latency(Latency<T>),
-    Jitter(Jitter<T>),
-    Rate(Rate<T>),
-    Seeking(Seeking<T>),
-    Segment(Segment<T>),
-    Convert(Convert<T>),
-    Formats(Formats<T>),
-    Buffering(Buffering<T>),
-    Custom(Custom<T>),
-    Uri(Uri<T>),
-    Allocation(Allocation<T>),
-    Scheduling(Scheduling<T>),
-    AcceptCaps(AcceptCaps<T>),
-    Caps(Caps<T>),
-    Drain(Drain<T>),
-    Context(Context<T>),
-    Other(Other<T>),
+#[derive(Debug)]
+pub enum QueryView {
+    Position(Position),
+    Duration(Duration),
+    Latency(Latency),
+    Jitter(Jitter),
+    Rate(Rate),
+    Seeking(Seeking),
+    Segment(Segment),
+    Convert(Convert),
+    Formats(Formats),
+    Buffering(Buffering),
+    Custom(Custom),
+    Uri(Uri),
+    Allocation(Allocation),
+    Scheduling(Scheduling),
+    AcceptCaps(AcceptCaps),
+    Caps(Caps),
+    Drain(Drain),
+    Context(Context),
+    Other(Other),
     __NonExhaustive,
 }
 
+impl Deref for QueryView {
+    type Target = Query;
+
+    fn deref(&self) -> &Self::Target {
+        match *self {
+            QueryView::Position(ref position) => position,
+            QueryView::Duration(ref duration) => duration,
+            QueryView::Latency(ref latency) => latency,
+            QueryView::Jitter(ref jitter) => jitter,
+            QueryView::Rate(ref rate) => rate,
+            QueryView::Seeking(ref seeking) => seeking,
+            QueryView::Segment(ref segment) => segment,
+            QueryView::Convert(ref convert) => convert,
+            QueryView::Formats(ref formats) => formats,
+            QueryView::Buffering(ref buffering) => buffering,
+            QueryView::Custom(ref custom) => custom,
+            QueryView::Uri(ref uri) => uri,
+            QueryView::Allocation(ref allocation) => allocation,
+            QueryView::Scheduling(ref scheduling) => scheduling,
+            QueryView::AcceptCaps(ref accept_caps) => accept_caps,
+            QueryView::Caps(ref caps) => caps,
+            QueryView::Drain(ref drain) => drain,
+            QueryView::Context(ref context) => context,
+            QueryView::Other(ref other) => other,
+            _ => unimplemented!("Converting QueryView into Query for unknown query type"),
+        }
+    }
+}
+
+// Bypass `QueryRc`: if `QueryView` is a `&mut`, this means that
+// `Query` can also be accessed as `&mut` thanks to the controls
+// enforced by `QueryWrapper`
+impl DerefMut for QueryView {
+    fn deref_mut(&mut self) -> &mut Query {
```

Comment:
I believe with `impl From<QueryRc> for QueryView` you can get a `QueryView` from every `GstRc<Query>`, and with the `deref_mut` here you could always get a `&mut Query`, no matter what the current refcount is

Reply:
That's true because `GstRc<Query>` is still in charge of creating the `QueryRef` from the `ffi::GstQuery`. But, if we consider the `QueryWrapper` is now in charge, we could avoid this. See below for the discussion about the `QueryWrapper` v.s. `GstRc<Query>`.

I will update the code to be more explcit.